src/base_obs.c

```diff
@@ -0,0 +1,329 @@
+           for smoothing. */
+        base_obss.pos_ecef[0] = 0.99995 * base_obss.pos_ecef[0]
+                                  + 0.00005 * soln.pos_ecef[0];
+        base_obss.pos_ecef[1] = 0.99995 * base_obss.pos_ecef[1]
+                                  + 0.00005 * soln.pos_ecef[1];
+        base_obss.pos_ecef[2] = 0.99995 * base_obss.pos_ecef[2]
+                                  + 0.00005 * soln.pos_ecef[2];
+      } else {
+        memcpy(base_obss.pos_ecef, soln.pos_ecef, 3 * sizeof(double));
+      }
+      base_obss.has_pos = 1;
+    } else {
+      /* There was an error calculating the position solution. */
+      printf("Error calculating base station position (%ld)\n", ret);
+    }
+  }
+  /* If the base station position is known then calculate the satellite ranges.
+   * This calculation will be used later by the propagation functions. */
+  if (base_obss.has_pos) {
+    for (u8 i=0; i < base_obss.n; i++) {
+      double dx[3];
+      vector_subtract(3, base_obss.nm[i].sat_pos, base_obss.pos_ecef, dx);
+      base_obss.sat_dists[i] = vector_norm(3, dx);
+    }
+  }
+  /* Unlock base_obss mutex. */
+  chMtxUnlock();
+  /* Signal that a complete base observation has been received. */
+  chBSemSignal(&base_obs_received);
+}
+
+/** SBP callback for observation messages.
+ * A set of observations may be split across several messages so this function
+ * attempts to collect a full set of observations into a single `obss_t`
+ * (`base_obss_raw`). Once a full set is received then update_obss() is called.
+ */
+static void obs_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void) context;
+
+  static s16 prev_count = 0;
+  static gps_time_t prev_t = {.tow = 0.0, .wn = 0};
+  /* Using an extra obss_t so we don't need to overwrite the old set (used for
+   * low latency) when we end up with a bad new set. */
+  static obss_t base_obss_raw = {.has_pos = 0};
+
+  /* Sender ID of zero means that the messages are relayed observations,
+   * ignore them. */
+  if (sender_id == 0)
+    return;
+
+  /* Relay observations using sender_id = 0. */
+  sbp_send_msg_(MSG_PACKED_OBS, len, msg, 0);
+
+  gps_time_t t;
+  u8 total;
+  u8 count;
+
+  /* Decode the message header to get the time and how far through the sequence
+   * we are. */
+  unpack_obs_header((msg_obs_header_t*)msg, &t, &total, &count);
+
+  /* Check to see if the observation is aligned with our internal observations,
+   * i.e. is it going to time match one of our local obs. */
+  double epoch_count = t.tow * (soln_freq / obs_output_divisor);
+  if (fabs(epoch_count - round(epoch_count)) > TIME_MATCH_THRESHOLD) {
+    printf("Unaligned observation from base station ignored.\n");
+    return;
+  }
+
+  /* Calculate packet latency. */
+  if (time_quality >= TIME_COARSE) {
+    gps_time_t now = get_current_time();
+    float latency_ms = (float) ((now.tow - t.tow) * 1000.0);
+    log_obs_latency(latency_ms);
+  }
+
+  /* Verify sequence integrity */
+  if (count == 0) {
```

Comment:
the semantics of count and prev_count are still unclear. Can we document what we're counting?